firebase-firestore/src/main/java/com/google/firebase/firestore/local/QueryEngine.java

```diff
@@ -110,17 +111,21 @@ public void initialize(LocalDocumentsView localDocumentsView, IndexManager index
       return null;
     }
 
-    List<DocumentKey> keys = indexManager.getDocumentsMatchingTarget(target);
-    if (keys == null) {
+    Pair<List<DocumentKey>, Boolean> indexResults = indexManager.getDocumentsMatchingTarget(target);
```

Comment:
Is it possible to handle all "usedFullIndex"/limit logic in QueryEngine? Prior to running the query we could possibly determine whether a full index is used and if so, we could drop the limit. 

Maybe we could have something like:

```
enum IndexStatus {
   None,
   Partial,
   Full
}

IndexStatus status = indexManager.canServerUsingIndex(target);

if (status.equals(None)) {
  return null;
}

if (status.equals(Partial)) { 
  // Witty yet descriptive comment
  query = query.limitToFirst(-1);
}
```

Reply:
What do you think?

Reply:
It's hard to say where to draw the line here. Your current approach has some benefits in it keeps the logic self contained for SQLite. We can add a different implementation (for the memory cache) for example that handles these queries differently. 
Adding this logic to the Query engine makes this part of the logic part of the main query evaluation path. It brings the concepts of "Partial" indexes from SQLite to the SDK as a hole and it creates a common way to execute limit queries with partial targets. I do think that even a memory-based indexing implementation will run into the same problem, and thus to me it makes sense to generalize this logic and have it in QueryEngine. But I cannot predict the future.

Reply:
I see your point. I've moved the logic to QueryEngine